src/mongocxx/private/database.hpp

```diff
@@ -34,6 +34,19 @@ class database::impl {
         name(std::move(name))
     {}
 
+    impl(const impl& i) : 
+        database_t{libmongoc::database_copy(i.database_t)},
+        client_impl{i.client_impl},
+        name{i.name} 
+    {}
+
+    impl& operator=(const impl& i) {
+        database_t = libmongoc::database_copy(i.database_t);
```

Comment:
I guess my interpretation of the rule of three (or rule of 5) isn't that you necessarily need to define all the lifecycle methods, but that if you define one, then you need to define (or delete) the rest as the compiler generated defaults will not be correct.

I think in this case we can just delete the copy assignment (and move assignment operator) of mongocxx::database::impl, and make the internal `database_t *` a `database_t* const`